### DISCUSSION TITLE
Support with Gherkin tests

### DISCUSSION THREAD

**Discussion by erdzan12** [leancodepl/patrol#1180] [Q&A]
Hi there,

Has anyone tried if Patrol can be used with the flutter_gherkin plugin?

**Comment by bartekpacia** [comment]
Hi @erdzan12 - yes, the awesome @nonvachara-paloit did this [here](https://github.com/leancodepl/patrol/discussions/652) :)

See also [this issue](https://github.com/leancodepl/patrol/issues/1160).

**Reply by erdzan12** [reply]
Thanks searched in the discussion with the filter closed and did not find it immediately

I see there the approach was to use the `patrol drive`  functionality which has been removed since `1.0.0`, and also other things changed as the `HostAutomator` and lack of the `patrolIntegrationDriver`. 
Any idea if now with the newest version, where we do not have the old functionality we would be able to do this with the flutter_gherkin plugin?

**Reply by bartekpacia** [reply]
> Any idea if now with the newest version, where we do not have the old functionality we would be able to do this with the flutter_gherkin plugin?

Yes, I'm pretty certain this is still possible.

We had quite a lot of changes in 1.0, so that code has to be adapted.

**Reply by erdzan12** [reply]
Great, got it to work with the android emulator.
Will provide in here some instructions on how I did it and also ask further questions, when I encounter problems.

**Reply by bartekpacia** [reply]
awesome! would be really great if you shared what you did here :)

**Reply by erdzan12** [reply]
I created here a repo with an example app using patrol and flutter_gherkin

I used here patrol to accept camera permissions.

https://github.com/erdzan12/patrol_gherkin

However, there is an issue there that, when starting the tests with 
```
patrol test --target integration_test/gherkin_suite_test.dart
```

It never ends the tests, it always shows 
```
Executing tests of apk with entrypoint gherkin_suite_test.dart on emulator-5554...
```
in the console.

**Reply by bartekpacia** [reply]
Thanks for research this and sharing with us!

As for the tests not finishing:

Please make sure you're using the newest `patrol` and `patrol_cli`. I see you're using `patrol: ^1.0.8`. Are you using the compatible Patrol CLI? Please see [release notes](https://pub.dev/packages/patrol_cli/changelog#114).

There were some changes under the hood in `patrol: 1.0.9` and `patrol_cli: 1.1.4`. General advice: _always_ use the latest versions.

**Reply by erdzan12** [reply]
I updated now the patrol versions, and both are the newest, but it still seems like not to finish

**Reply by erdzan12** [reply]
The last item when running the tests with verbose is:
```
        : > Task :app:connectedDebugAndroidTest
```
I'm not sure if there might something special which should be returned at the end of the tests to let patrol know they are done

Which is called after patrol does this:

```
./gradlew :app:connectedDebugAndroidTest -Ptarget=/Users/user/Documents/temp_flutter_projects/patrol_gherkin/integration_test/gherkin_suite_test.dart -Pdart-defines=UEFUUk9MX1dBSVQ9MA==,UEFUUk9MX0FQUF9QQUNLQUdFX05BTUU9Y29tLmV4YW1wbGUucGF0cm9sX2doZXJraW4=,UEFUUk9MX0FORFJPSURfQVBQX05BTUU9UGF0cm9sIEdoZXJraW4gVGVzdCBBcHA=,UEFUUk9MX0lPU19BUFBfTkFNRT1QYXRyb2wgR2hlcmtpbiBUZXN0IEFwcA==,SU5URUdSQVRJT05fVEVTVF9TSE9VTERfUkVQT1JUX1JFU1VMVFNfVE9fTkFUSVZFPWZhbHNl,UEFUUk9MX1RFU1RfTEFCRUw9Z2hlcmtpbl9zdWl0ZV90ZXN0LmRhcnQ=
```

**Reply by bartekpacia** [reply]
Take a look at [PatrolBinding](https://github.com/leancodepl/patrol/blob/371356fe84db6ce78d97ec181c14fd85027154cb/packages/patrol/lib/src/binding.dart). It collects Dart test results as they execute, and once they finish, it sends these results over gRPC to the _test app_. Patrol CLI should exit once the results arrive in the test app.

See also:
- [PatrolTestRunner.kt](https://github.com/leancodepl/patrol/blob/371356fe84db6ce78d97ec181c14fd85027154cb/packages/patrol/android/src/main/kotlin/pl/leancode/patrol/PatrolTestRunner.kt#L62)

**Reply by erdzan12** [reply]
Ok i see now the problem, the gherkin tests are using `GherkinIntegrationTestRunner` which is using the `IntegrationTestWidgetsFlutterBinding` and not the `PatrolBinding`, I changed it in the source code of flutter_gherkin to use the `PatrolBinding`, so it seems like I would have to create a PR in the flutter_gherkin package to accept custom bindings extending the `IntegrationTestWidgetsFlutterBinding `
Or do you see if it could work without the `PatrolBinding`?

**Reply by bartekpacia** [reply]
Yeah, looks like support for custom bindings has to be added to flutter_gherkin.

**Reply by creatorpiyush** [reply]
Hello @erdzan12, can you please share the PR/Branch you raised for flutter_gherkin for custom binding support
